### PR TITLE
Enable mutating content type from proxies

### LINKED ISSUE #298: Proxies don't support changing the type of an asset
Proxies don't support changing the type of an asset as `Sprockets.register_transformer` does. As far as I can tell, there are currently two technical reasons we currently don't have this support but they are easy to rectify and **I've nearly completed a patch.**

The first reason is that there is simply no API to do so. Firstly we must establish whether to deal in file extension or mime type. I suggest mime type and `ProxiedAsset#content_type=`.

The second reason is that Jekyll Asset's internal cache does not support changing the extension. This is easily rectified by looking up only the `basename` of the cached filename using `Dir.glob`. This also means, after a proxy has been called, if it used `ProxiedAsset#content_type=`, the old cached file will be deleted. There is no reason to keep this file around since we cache based on a digest of the proxies used and it will only confuse the cache and take up space.

Thoughts?

/cc @envygeeks

### PR DESCRIPTION
**Pull request by spudowiar**
Fixes #298

### CODE REVIEW

File: lib/jekyll/assets/liquid/tag/proxied_asset.rb [hunk 1]
```diff
@@ -77,6 +77,19 @@ def digest
           end
 
           # ------------------------------------------------------------------
+
+          attr_reader :content_type
```
Comment (envygeeks) [inline lib/jekyll/assets/liquid/tag/proxied_asset.rb]: This belongs at the top of the class with the delegators.

File: lib/jekyll/assets/liquid/tag/proxied_asset.rb [hunk 2]
```diff
@@ -77,6 +77,19 @@ def digest
...
+
+          def content_type=(content_type)
+            if @content_type != content_type
```
Comment (envygeeks) [inline lib/jekyll/assets/liquid/tag/proxied_asset.rb]: `return unless @content_type != content_type`

File: lib/jekyll/assets/liquid/tag/proxied_asset.rb [hunk 3]
```diff
@@ -94,13 +107,30 @@ def digest_path
...
           # ------------------------------------------------------------------
 
           private
           def proxy_file
             unless cached?
               args.proxies.each do |key, val|
+                old_type = content_type
+                old_filename = filename
                 Proxies.get(key).first[:class].new(self, val, @args).process
+                File.delete old_filename if old_type != content_type
```
Comment (envygeeks) [inline lib/jekyll/assets/liquid/tag/proxied_asset.rb]: This should use `Pathutil` not file. and filename should already be an instance of Pathutil.

File: lib/jekyll/assets/liquid/tag/proxied_asset.rb [hunk 4]
```diff
@@ -109,8 +139,12 @@ def proxy_file
 
           private
           def cache_file
-            @_cached = File.file?(filename)
-            return @_cached if @_cached
+            glob = Dir.glob("#{filename.chomp(File.extname(filename))}.*")
```
Comment (envygeeks) [inline lib/jekyll/assets/liquid/tag/proxied_asset.rb]: * If filename is just a filename: `Pathutil.new(".").glob(Pathutil.new(filename).sub_ext(".*"))`
* If filename is just a absolute: `filename.dirname.glob(filename.basename.sub_ext(".*"))`

File: lib/jekyll/assets/liquid/tag/proxied_asset.rb [hunk 5]
```diff
@@ -109,8 +139,12 @@ def proxy_file
...
+            @_cached = !glob.empty?
+            if @_cached
+              self.content_type = _mime_for(File.extname(glob.first))
```
Comment (envygeeks) [inline lib/jekyll/assets/liquid/tag/proxied_asset.rb]: When you move to Pathutil you'll move this to `glob.first.extname` only.  Skipping `File`

File: lib/jekyll/assets/liquid/tag/proxied_asset.rb [hunk 6]
```diff
@@ -94,13 +107,30 @@ def digest_path
             "#{name.chomp(ext)}-#{digest}#{ext}"
           end
 
+          # --------------------------------------------------------------------
+
+          private
+          def _mime_for(ext)
+            Sprockets.mime_exts[ext]
```
Comment (envygeeks) [inline lib/jekyll/assets/liquid/tag/proxied_asset.rb]:
```
[
  ext
]
```

**Comment by envygeeks** [comment]
Looking good, thanks! Just a few comments.

File: lib/jekyll/assets/liquid/tag/proxied_asset.rb [hunk 7]
```diff
@@ -16,18 +16,20 @@ class ProxiedAsset
...
-          def_delegator :@asset, :content_type
+          def_delegator :@asset, :source_filename
           def_delegator :@asset, :mtime
 
           # ------------------------------------------------------------------
 
           def initialize(asset, args, env, tag)
             @env = env
-            @asset = asset
+            @asset = asset.dup
```
Comment (envygeeks) [inline lib/jekyll/assets/liquid/tag/proxied_asset.rb]: What is the reason for duping the object?

File: lib/jekyll/assets/liquid/tag/proxied_asset.rb [hunk 8]
```diff
@@ -59,13 +61,13 @@ def write_to(name)
...
           def source
-            File.binread(filename)
+            filename.binread
           end
 
           # ------------------------------------------------------------------
 
           def filename
-            env.in_cache_dir(digest_path)
+            Pathutil.new env.in_cache_dir(digest_path)
```
Comment (envygeeks) [inline lib/jekyll/assets/liquid/tag/proxied_asset.rb]:
```
Pathutil.new(
  // Stuff
)
```

File: lib/jekyll/assets/liquid/tag/proxied_asset.rb [hunk 9]
```diff
@@ -77,6 +79,16 @@ def digest
           end
 
           # ------------------------------------------------------------------
+
+          def content_type=(type)
+            return if @content_type == type
+
+            path = Pathutil.new asset.logical_path
```
Comment (envygeeks) [inline lib/jekyll/assets/liquid/tag/proxied_asset.rb]: Methods that have arguments should have `()` for clarity.  Unless it's a non-sub-local `asset arg` would be fine (I would let it slip) but `asset.blah val` will probably end up getting corrected later so it's clearer.

**Comment by envygeeks** [comment]
Just a few more comments and one note: this will need tests to regression proof your feature because there will be heavy rewriting come Sprockets 4, I can already feel it by the amount of smoke that came up when I threw some water on it a few weeks.

**Comment by spudowiar** [comment]
Yep, I'm totally aware of that and they are in the works. I just wanted to get my initial code up for you to 'lint' :)

File: lib/jekyll/assets/liquid/tag/proxied_asset.rb (hunk 7 above)
Comment (spudowiar) [inline lib/jekyll/assets/liquid/tag/proxied_asset.rb]: > cdf6de7266f1da9cfa9662b9e9bd29815a7b1270 Proxies: fix mutable content type affecting later use of asset

Changing the content type once will affect it in the future. I'm just fixing up an alternative to this which isn't so expensive.

**Comment by spudowiar** [comment]
Right, just tests to write now!

**Comment by envygeeks** [comment]
Looks like you broke your own pull request, gotta rebase when you throw up the specs.

**Comment by spudowiar** [comment]
@envygeeks check it over, but it's ready to merge AFAICT :shipit:

**Comment by envygeeks** [comment]
Hulk: +security
@spudowiar can you rebase this into logical commits.

**Comment by spudowiar** [comment]
@envygeeks sorry for leaving this so long, better late than never though!

**Comment by envygeeks** [comment]
:heart:

**Pull request merged**